### PR TITLE
test: add integration test for flink shell

### PR DESCRIPTION
**Pull request by YannickPferr**
Release Notes
-------------

Checklist
---------
- [x] Leave this box unchecked if features are not yet available in production

What
----

Added integration tests for the flink shell. I tried a bunch of different methods, but because go-prompt is interacting with `/dev/tty` directly, it is quite hard to submit commands programatically to the shell. T. work around this, we now create a file that will then be used as the input stream for go-prompt.

References
----------

Test & Review
-------------

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -25,6 +25,7 @@ blocks:
             - make generate-packaging-patch
             - diff -w -u <(git cat-file --filters HEAD:debian/patches/standard_build_layout.patch | awk "{if (NR>3) {print}}") <(cat debian/patches/standard_build_layout.patch | awk "{if (NR>3) {print}}")
             - make lint
+            - go install github.com/charmbracelet/vhs@latest
```
Comment (brianstrauch) [inline .semaphore/semaphore.yml:28]: :heart:

**Review by Guttz — APPROVED** [review]

File: test/fixtures/input/flink/flinkshell-commands [hunk 2]
```diff
@@ -0,0 +1,11 @@
+use catalog default;
```
Comment (brianstrauch) [inline test/fixtures/input/flink/flinkshell-commands:1]: Can integration tests test only one (or if necessary, two) statement at a time. From previous experience, once you start chaining multiple statements/commands in each test it gets harder to debug.

File: test/fixtures/output/flink/shell/shell.golden [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+TEXT OUTPUT 
+========================================================================== 
+
+Welcome! 
+To exit, press Ctrl-Q or type "exit". 
+
+[Ctrl-Q] Quit [Ctrl-S] Toggle Smart Completion 
+ 
+Smart Completion is now disabled                                                                    
+> > > use catalog default;> use catalog default;> use catalog default;
```
Comment (brianstrauch) [inline test/fixtures/output/flink/shell/shell.golden:11]: Looks like we're submitting multiple statements in parallel, without waiting for the previous one to execute?

File: test/fixtures/output/flink/shell/shell.golden [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
++------------------------+---------------------+
+|          Key           |        Value        |
++------------------------+---------------------+
+| client.service-account | <unset> (default)   |
+| sql.local-time-zone    | GMT+00:00 (default) |
++------------------------+---------------------+
+ > exit> exit> exit
+]2;
+
+TEXT + ANSI ESCAPE SEQUENCES OUTPUT
```
Comment (brianstrauch) [inline test/fixtures/output/flink/shell/shell.golden:118]: Is this really that useful? Personally, I just find it really confusing. If we make a PR that changes the output colors, wouldn't it be easier to manually verify the output?

File: test/flink_test.go [hunk 5]
```diff
@@ -120,3 +142,139 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+
+	var outputBuffer bytes.Buffer
+	cmd.Stdout = &outputBuffer
+	cmd.Stderr = &outputBuffer
+
+	err = cmd.Start()
+	require.NoError(s.T(), err, outputBuffer.String())
+
+	// wait for flink shell to be ready
+	time.Sleep(5 * time.Second)
```
Comment (brianstrauch) [inline test/flink_test.go:179]: Is there a better way to verify this?

File: test/flink_test.go [hunk 6]
```diff
@@ -120,3 +142,139 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
 		s.runIntegrationTest(test)
 	}
 }
+
+func (s *CLITestSuite) TestFlinkShell() {
```
Comment (brianstrauch) [inline test/flink_test.go:146]: Is the idea to replace all existing `flink shell` integration tests with this style of testing?

File: test/fixtures/output/flink/shell/shell.golden (hunk 3 above)
Comment (YannickPferr) [reply test/fixtures/output/flink/shell/shell.golden:11]: No I think this just comes from @Guttz removing the ansi sequences to make the golden file more readable

File: test/fixtures/output/flink/shell/shell.golden (hunk 4 above)
Comment (YannickPferr) [reply test/fixtures/output/flink/shell/shell.golden:118]: hmm yeah I kinda agree, I was also considering disabling colors since it will make the golden file easier to read

File: test/flink_test.go (hunk 5 above)
Comment (YannickPferr) [reply test/flink_test.go:179]: hmm maybe.. Perhaps we could wait for the `Welcome` text? 🤔

File: test/flink_test.go (hunk 6 above)
Comment (YannickPferr) [reply test/flink_test.go:146]: I think it still makes sense to keep our existing more unit style tests since they are fast, easier to implement and enable us to test certain edge cases more easily. This is just to also have a test for actually starting up the binary and executing some commands without mocking everything.

Currently we're also missing a mock backend and just test local commands, I was planning to add a mock backend in the future to fake for example a `select` statement

File: test/fixtures/input/flink/flinkshell-commands (hunk 2 above)
Comment (YannickPferr) [reply test/fixtures/input/flink/flinkshell-commands:1]: yeah we can split it up a bit, I was thinking that it is faster this way, since we don't have to do the setup all over again.

File: test/fixtures/output/flink/shell/shell.golden (hunk 3 above)
Comment (brianstrauch) [reply test/fixtures/output/flink/shell/shell.golden:11]: Any way we can clean up the formatting here?

Comment (YannickPferr) [reply test/fixtures/output/flink/shell/shell.golden:11]: I'll disable the colors and then there should be no need for removing the ansi sequences, which will in turn make the golden file more beautiful to look at 😄

File: test/flink_test.go (hunk 6 above)
Comment (brianstrauch) [reply test/flink_test.go:146]: > Currently we're also missing a mock backend and just test local commands, I was planning to add a mock backend in the future to fake for example a select statement

This would be great. The end goal would be for our existing integration tests and the Flink integration tests to use the same framework... although the DevProd team has been trying to switch us over to contract tests lately.

**Comment by YannickPferr** [comment]
Also just noticed the race in CI, will fix this and then try to address the rest of the things tomorrow

File: test/flink_test.go [hunk 7]
```diff
@@ -120,3 +152,214 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+	tests := []FlinkShellTest{
```
Comment (Guttz) [inline test/flink_test.go:157]: nit: wdyt of listing tests from folder to exclude one step when adding test?


```
    files, err := os.ReadDir('input/flink/shell')
    if err != nil {
        fmt.Println("Error reading directory:", err)
        return
    }

    for _, file := range files {
         testName := file.Name()
         s.runFlinkShellTest({{inputFile: testName, goldenFile: testName + ".golden" }})
    }
```

File: test/flink_test.go [hunk 8]
```diff
@@ -120,3 +152,214 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+		s.runFlinkShellTest(test)
+	}
+}
+
+func (s *CLITestSuite) runFlinkShellTest(flinkShellTest FlinkShellTest) {
+	s.T().Run(flinkShellTest.inputFile, func(t *testing.T) {
+		s.login(t)
+
+		// Create a file for go-prompt to use as the input stream
+		stdin, err := os.Create("input.txt")
```
Comment (Guttz) [inline test/flink_test.go:175]:
```suggestion
		stdin, err := os.Create("flink_shell_input_stream.txt")
```
maybe reduce the chance that the CI starts to fail because someone created a input.txt file here? lol

File: test/flink_test.go [hunk 9]
```diff
@@ -120,3 +152,214 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+		// Register stdout scanner
+		pipe, err := cmd.StdoutPipe()
+		require.NoError(t, err)
+		defer pipe.Close()
+		stdoutScanner := bufio.NewScanner(pipe)
+
+		// Start command
+		err = cmd.Start()
+		require.NoError(t, err)
+		go killCommandIfTimeoutExceeded(flinkShellTest.timeout, cmd)
```
Comment (Guttz) [inline test/flink_test.go:197]: nice

File: test/flink_test.go [hunk 10]
```diff
@@ -120,3 +152,214 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+	time.Sleep(timeout)
+	if err := cmd.Process.Kill(); err != nil {
+		fmt.Printf("failed to kill command %v/n", err)
+	}
+}
+
+func waitForFlinkShellToBeReady(scanner *bufio.Scanner, output *strings.Builder) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		output.WriteString(line + "\n")
```
Comment (Guttz) [inline test/flink_test.go:293]: do we have to execute this for every scan before checking if strings.Contains(line, "[Ctrl-Q]..?

**Review by Guttz — COMMENTED** [review]
Beautiful 💯

File: test/flink_test.go (hunk 7 above)
Comment (brianstrauch) [reply test/flink_test.go:157]: Or better yet, we could just use a `[]string` instead of a file for the input? A good test will have a short script of statements so this should be fine.

File: test/flink_test.go [hunk 11]
```diff
@@ -120,3 +152,214 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func removeANSIEscapeSequences(input string) string {
```
Comment (brianstrauch) [inline test/flink_test.go:341]:
```suggestion
func removeAnsiEscapeSequences(input string) string {
```

Comment (YannickPferr) [reply test/flink_test.go]: renamed!

File: test/flink_test.go (hunk 7 above)
Comment (YannickPferr) [reply test/flink_test.go:157]: yeah sounds good, I wanted to make use of the input folder 😄 Using a string array now instead!

File: test/flink_test.go [hunk 12]
```diff
@@ -155,11 +155,48 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+			name:       "TestUseDatabase",
+			goldenFile: "use-database.golden",
+			commands: []string{
+				"use db1;",
+				"set;",
+			},
+		},
+		{
+			name:       "TestSetSingleKey",
+			goldenFile: "set-key.golden",
```
Comment (brianstrauch) [inline test/flink_test.go:176]: Another great improvement might be to let the golden files describe the test, instead of `name`. This'll make it much quicker to write new tests.
```suggestion
			goldenFile: "set-single-key.golden",
```

File: test/flink_test.go [hunk 13]
```diff
@@ -120,3 +152,244 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+				"set;",
+			},
+		},
+		{
+			name:       "TestResetAllKeys",
+			goldenFile: "reset-all.golden",
+			commands: []string{
+				"set 'cli.a-key'='a value';",
+				"set 'cli.another-key'='another value';",
+				"reset;", "set;",
```
Comment (brianstrauch) [inline test/flink_test.go:197]:
```suggestion
				"reset;",
				"set;",
```

File: test/flink_test.go [hunk 14]
```diff
@@ -120,3 +152,244 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+	input = stripCursorUp.ReplaceAllString(input, "")
+
+	return strings.TrimSpace(input)
+}
+
+func submitStatement(stdin *os.File) error {
+	// We need to wait before and after writing a new line, so go-prompt receives the ENTER as an individual key press
+	time.Sleep(50 * time.Millisecond)
+	_, err := stdin.WriteString("\n")
+	time.Sleep(50 * time.Millisecond)
```
Comment (brianstrauch) [inline test/flink_test.go:393]: I see a few `Sleep()` functions hidden throughout these tests. How long do these tests take end-to-end? Can we decrease this to 10ms? Can we scan for specific text to know when we can write a newline?

File: test/flink_test.go [hunk 15]
```diff
@@ -120,3 +152,244 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+	return err
```
Comment (brianstrauch) [inline test/flink_test.go:394]: If there was an error, wouldn't we want to return it right away instead of waiting 50ms?

File: test/flink_test.go (hunk 14 above)
Comment (YannickPferr) [reply test/flink_test.go:393]: hmm let me see, maybe we can wait for the `Statement successfully submitted.` text 🤔

File: test/flink_test.go (hunk 10 above)
Comment (Guttz) [reply test/flink_test.go:329]: Ah, I understand! You do this to print whatever we have as welcome message/headers when starting the CLI.

File: test/flink_test.go (hunk 14 above)
Comment (YannickPferr) [reply test/flink_test.go]: okay that works, added a helper that waits for a particular line to appear and then returns the output recorded up until that point

File: test/flink_test.go (hunk 15 above)
Comment (YannickPferr) [reply test/flink_test.go]: should be resolved as well with the latest commit

File: test/flink_test.go (hunk 12 above)
Comment (YannickPferr) [reply test/flink_test.go]: done

File: test/flink_test.go (hunk 13 above)
Comment (YannickPferr) [reply test/flink_test.go]: fixed

File: test/flink_test.go (hunk 14 above)
Comment (YannickPferr) [reply test/flink_test.go]: > How long do these tests take end-to-end?

oh and the tests take about 7s to run locally atm:
```
--- PASS: TestCLI (6.73s)
    --- PASS: TestCLI/TestFlinkShell (4.34s)
        --- PASS: TestCLI/TestFlinkShell/use-catalog (1.67s)
        --- PASS: TestCLI/TestFlinkShell/use-database (0.81s)
        --- PASS: TestCLI/TestFlinkShell/set-single-key (0.54s)
        --- PASS: TestCLI/TestFlinkShell/reset-single-key (0.61s)
        --- PASS: TestCLI/TestFlinkShell/reset-all-keys (0.73s)

```

Comment (brianstrauch) [reply test/flink_test.go]: Huh, I wonder if that's including the time to build the binary? Let's see where else we can speed up these tests... 1s per integration test is not going to scale well

Comment (Guttz) [reply test/flink_test.go]: I think we could create another ticket on how to increase the performance here, wdyt @brianstrauch? Yannick has been working on this over a week, and the whole ticket already required much more time than we planned

Comment (Guttz) [reply test/flink_test.go]: Also, I don't think we're planning to add 100s of integration tests now. However, if we start adding more we'll prioritize increasing the performance first

Comment (YannickPferr) [reply test/flink_test.go]: on CI the test takes about 2s, which seems reasonable to me 🤔  
image

File: test/flink_test.go [hunk 16]
```diff
@@ -120,3 +151,233 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+		s.login(t)
+
+		// Create a file for go-prompt to use as the input stream
+		stdin, err := os.Create("flink_shell_input_stream.txt")
+		require.NoError(t, err, "error creating file")
+		defer cleanupInputFile(stdin)
+
+		envVars, err := setEnvVars(stdin.Name())
+		require.NoError(t, err, "error setting env vars")
+		defer cleanupEnvVars(envVars)
```
[the remark this answers is not in the document]
Comment (YannickPferr) [reply test/flink_test.go:214]: yeah let me do that

File: test/flink_test.go [hunk 17]
```diff
@@ -120,3 +151,233 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
...
+			}),
+		)
+		assert.NoError(t, snapshotConfig.SnapshotWithName(flinkShellTest.goldenFile, output.String()),
+			fmt.Sprintf("full output was %s", output.String()))
+	})
+}
+
+func (s *CLITestSuite) login(t *testing.T) {
+	loginString := fmt.Sprintf("login --url %s", s.TestBackend.GetCloudUrl())
+	env := []string{pauth.ConfluentCloudEmail + "=fake@user.com", pauth.ConfluentCloudPassword + "=pass1"}
```
[the remark this answers is not in the document]
Comment (YannickPferr) [reply test/flink_test.go:260]: I could also hardcode them, but the runCommand func needs an array and does some things with it: https://github.com/confluentinc/cli/blob/6d0ca9fad8d79cf4664dd8ef9880adc2aeb9838f/test/cli_test.go#L225 

Maybe I could pass an empty array here?

File: test/flink_test.go [hunk 18]
```diff
@@ -1,5 +1,37 @@
...
+const (
+	flinkShellInputStreamFile     = "flink_shell_input_stream.txt"
+	flinkShellFixtureOutputFolder = "test/fixtures/output/flink/shell"
+	timezoneEnvVar                = "TZ"
+)
+
+type flinkShellTest struct {
+	commands   []string
+	goldenFile string
+	timeout    time.Duration
```
Comment (brianstrauch) [inline test/flink_test.go:32]: This looks unused. Can we leave it out until we need it?

File: test/flink_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (brianstrauch) [reply test/flink_test.go]: Oops, I didn't see that. It looks like we're setting these environment variables twice, so let's change that to only be once.

**Review by brianstrauch — APPROVED** [review]

**Review by Guttz — APPROVED** [review]
